### ISSUE TITLE
Adapter to OrientDB

### ISSUE THREAD

**Issue by niphlod** [web2py/web2py#603]
_From [yanosh...@gmail.com](https://code.google.com/u/102304726775744565779/) on September 03, 2011 20:23:02_

What steps will reproduce the problem? 1. 2. 3. What is the expected output? What do you see instead? What version of the product are you using? On what operating system? Please provide any additional information below. I wish to collaborate on/test one as I am not a super Python pro yet.  This Document-Graph database has a SQL syntax, RESTful interface, and the Gremlin graph traversal language bundled into it.

That brings untold power to web2py to store/retrieve schema-less, schema-mixed and schema-full data which can be traversed orders of magnitude faster than traditional RDBMS.

ACID, transactions, network pooling, more that I haven't discovered yet.

Thank you all for your work.  web2py is the best and adding OrientDB communication to it would make heads explode across the globe.

_Original issue: http://code.google.com/p/web2py/issues/detail?id=407_

**Comment by niphlod** [comment]
_From [yanosh...@gmail.com](https://code.google.com/u/102304726775744565779/) on September 16, 2011 11:40:51_

-David Bloom

**Comment by niphlod** [comment]
_From [massimo....@gmail.com](https://code.google.com/u/111721297562664670951/) on December 16, 2011 08:26:59_

Accepted. Which of the current adapters is the closest to the OrientDB syntax (sqlite, myqsl, postgresql, mssql?)

**Status:** Accepted

**Comment by niphlod** [comment]
_From [yanosh...@gmail.com](https://code.google.com/u/102304726775744565779/) on December 17, 2011 08:16:37_

The rest of the SQL syntax is very much like MySQL/Postgres.  See: https://code.google.com/p/orient/wiki/SQL I see a potential problem in that OrientDB uses strings for its id field. I haven't looked recently but as I understand it the dbs in dal.py use integers as id.  The OrientDB wiki will explain it better than I:

"RecordID

In OrientDB each record has a unique ID. The RecordID is composed in this way: #[<cluster>:<position>] Where:

    cluster, is the cluster id. Positive numbers mean physical clusters. Negative numbers mean logical clusters.
    position, is the absolute position of the record inside a cluster. 

NOTE: After the release 1.0rc4 the prefix character # is mandatory to recognize a RecordID.

The record never looses its identity unless is deleted. Once deleted its identity could be recycled and assigned to a new record. See the Inverse relationships to know more about this.

You can access directly to a record by knowing its RecordID. For this reason you don't need to create a field as primary key as happens in Relational DBMS. "


I hope the string as an ID issue can be overcome.

Not sure if it will help but I have had success in the meantime using the requests and ujson python libs to make HTTP requests to OrientDB's RESTful API and then .loads() the JSON response.  There is a network binary protocol but I don't yet know how to have web2py and Orient communicate via binary.

It is an exciting prospect to combine web2py's flexibility with OrientDB's speed and analytical power.  I think it would be amazing to not only have an adapter to web2py but also to have the communications somehow include running Gremlin graph traversals too.

Thanks again Massimo.  I saw you so busy with the book and new versions and didn't think you'd get to my issue so soon.  Do you ever sleep?  :)

-David

**Comment by niphlod** [comment]
_From [massimo....@gmail.com](https://code.google.com/u/111721297562664670951/) on December 17, 2011 09:21:33_

Can you point me to the function to obtain a new record id after a new record is inserted?  
\> Do you ever sleep? 
I did yesterday but I do not remember the previous time I did before yesteray. ;-)

Massimo

**Comment by niphlod** [comment]
_From [yanosh...@gmail.com](https://code.google.com/u/102304726775744565779/) on December 17, 2011 10:10:25_

Ha, glad you got at least one night's rest.

I wrote a little module which I used to parse the RID from the JSON result
of an INSERT.

Works great in /modules.  Not sure if I need the from gluon import * bit
but left it as it was there when I created a new module.

/modules/orientdb_utils.py
\------------------------------------------
\#!/usr/bin/env python
\# coding: utf8
from gluon import *
import re

def RidFromInsert(resultstr):
    regxrslt = re.search( '#\d+:\d+', resultstr )
    return regxrslt.group(0)
\------------------------------------------

Example of using requests, ujson and module to insert a record and retrieve
it's new RID.
This assumes a running OrientDB instance on the localhost with a database
called "web2pyrules" and has a schema that has a cluster called Category
Think of OrientDB cluster = MySQL table.  A way to group information except
it can be schema-full/mixed/less


import requests
import ujson as json    # Faster JSON
from orientdb_utils import RidFromInsert

r = requests.post('http://localhost:2480/command/web2pyrules/sql'
data="INSERT INTO Category (name) VALUES ('web2py')", auth=('writer',
'writer'))
result = json.loads(r.content)
cat_id = RidFromInsert(result)


I don't know if it would be useful but I also wrote a module to add/remove
in/out edges to a vertex when using OrientDB as a document-graph db.  It
also uses the RESTful API similar to above.
